tools/jnimarshalmethod-gen/App.cs

```diff
@@ -465,17 +468,70 @@ void CreateMarshalMethodAssembly (string path)
 			typeof (string),
 		});
 
-		static Expression CreateRegistration (string method, string signature, LambdaExpression lambda, ParameterExpression targetType, string methodName)
+		static void CreateDelegateRuntimeManagedMethod (TypeBuilder tb, string name, Type returnType, Type[] parameterTypes)
+		{
+			var mb = tb.DefineMethod (name,
+				System.Reflection.MethodAttributes.Public |
+				System.Reflection.MethodAttributes.HideBySig |
+				System.Reflection.MethodAttributes.NewSlot |
+				System.Reflection.MethodAttributes.Virtual,
+				CallingConventions.Standard, returnType, parameterTypes);
+			mb.SetImplementationFlags (System.Reflection.MethodImplAttributes.Runtime | System.Reflection.MethodImplAttributes.Managed);
+		}
+
+		static Expression CreateRegistration (string method, string signature, LambdaExpression lambda, ParameterExpression targetType, string methodName, ModuleBuilder dm, List<Type> createdDelegateList)
 		{
 			Expression registrationDelegateType = null;
 			if (lambda.Type.Assembly == typeof (object).Assembly ||
 					lambda.Type.Assembly == typeof (System.Linq.Enumerable).Assembly) {
 				registrationDelegateType = Expression.Constant (lambda.Type, typeof (Type));
 			}
 			else {
+				string delegateTypeName;
+				if (lambda.Type.Assembly.IsDynamic) {
+					var typeNameBuilder = new StringBuilder ("__<$>_jni_marshal_");
+					var parameterTypes = new List<Type> ();
+					foreach (var p in lambda.Parameters) {
+						parameterTypes.Add (p.Type);
+					}
+
+					MarshalMemberBuilder.AddMarshalerTypeNameSuffix (typeNameBuilder, lambda.ReturnType, parameterTypes);
+
+					var typeName = typeNameBuilder.ToString ();
+					var existingType = dm.GetType (typeName);
+					if (existingType == null) {
+						var dtb = dm.DefineType (typeName,
+							System.Reflection.TypeAttributes.AnsiClass |
+							System.Reflection.TypeAttributes.Sealed,
+							typeof (MulticastDelegate));
+
+						var dc = dtb.DefineConstructor (
+							System.Reflection.MethodAttributes.Public |
+							System.Reflection.MethodAttributes.HideBySig |
+							System.Reflection.MethodAttributes.RTSpecialName |
+							System.Reflection.MethodAttributes.SpecialName,
+							CallingConventions.Standard, new Type [] { typeof (object), typeof (IntPtr) });
+						dc.SetImplementationFlags (System.Reflection.MethodImplAttributes.Runtime | System.Reflection.MethodImplAttributes.Managed);
+
+						CreateDelegateRuntimeManagedMethod (dtb, "Invoke", typeof (bool), parameterTypes.ToArray ());
+
+						parameterTypes.Add (typeof (AsyncCallback));
+						parameterTypes.Add (typeof (object));
+
+						CreateDelegateRuntimeManagedMethod (dtb, "BeginInvoke", typeof (IAsyncResult), parameterTypes.ToArray ());
```

Comment:
Do we need to declare `BeginInvoke()` and `EndInvoke()` if these are generated types, with type names that are "inaccessible", and `jnimarshalmethod-gen` *isn't* emitting references to these methods?